pkg/ruler/ruler.go

```diff
@@ -368,6 +297,143 @@ func (r *Ruler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 				</body>
 			</html>`
 		w.WriteHeader(http.StatusOK)
-		w.Write([]byte(unshardedPage))
+		_, err := w.Write([]byte(unshardedPage))
+		if err != nil {
+			level.Error(util.Logger).Log("msg", "unable to serve status page", "err", err)
+		}
+	}
+}
+
+func (r *Ruler) run() {
+	defer close(r.terminated)
+
+	tick := time.NewTicker(r.cfg.PollInterval)
+	defer tick.Stop()
+
+	r.loadRules()
+	for {
+		select {
+		case <-r.done:
+			return
+		case <-tick.C:
+			r.loadRules()
+		}
+	}
+}
+
+func (r *Ruler) loadRules() {
+	ringHasher := fnv.New32a()
+
+	configs, err := r.store.ListAllRuleGroups(context.Background())
```

Comment:
`loadRules()` could create a context at the top and pass it into `newManager()` so that the same context is used in both places since this functionality is connected.  unsure if that would connect traces or do anything valuable.

Reply:
👍 I'm not sure there is any benefits currently but down the line we could consider using a context for top level cancellation.